jerry-core/ecma/base/ecma-helpers-string.c

```diff
@@ -198,6 +203,26 @@ ecma_new_ecma_string_from_magic_string_ex_id (lit_magic_string_ex_id_t id) /**<
   return string_desc_p;
 } /* ecma_new_ecma_string_from_magic_string_ex_id */
 
+#ifndef CONFIG_DISABLE_ES2015_SYMBOL_BUILTIN
+/**
+ * Allocate new ecma-string and fill it with reference to the symbol descriptor
+ *
+ * @return pointer to ecma-string descriptor
+ */
+ecma_string_t *
+ecma_new_symbol_from_descriptor_string (ecma_value_t string_desc) /**< ecma-string */
+{
+  JERRY_ASSERT (!ecma_is_value_symbol (string_desc));
+
+  ecma_string_t *symbol_p = ecma_alloc_string ();
+  symbol_p->refs_and_container = ECMA_STRING_REF_ONE | ECMA_STRING_CONTAINER_HEAP_UTF8_SYMBOL;
+  symbol_p->u.symbol_descriptor = string_desc;
+  symbol_p->hash = string_desc & ECMA_SYMBOL_HASH_MASK;
```

Comment:
Could symbols with identical descriptions being treated as unique keys be achieved by performing a bit-and an incremental counter here?

Reply:
Not quite so. Each symbol is unique since we allocate a new string for it. The address of the new string will be the unique key. The hash property only stands for performance reasons.

Reply:
> Each symbol is unique since we allocate a new string for it.

I'm wondering if JerryScript would reuse string value `'foo'` in following case?

```js
var symbol1 = Symbol('foo')
var symbol2 = Symbol('foo')
```

If it does, then both `symbol1` and `symbol2` has same hash value, which ended with same property entry:

```js
var obj = {}
obj[symbol1] = 'foo'
obj[symbol2] = 'bar' // overrides obj[symbol1]
```

Reply:
1. Yes, the engine reuses the `foo`. 
`symbol1` is a new `ecma-string` which container is `ECMA_STRING_CONTAINER_HEAP_UTF8_SYMBOL`, and the `foo` description string is stored in the `symbol_p->u.symbol_descriptor`. So it only puts and extra reference to `foo`.

In case of `symbol2` exactly happens the same, so `foo` now has +2 extra reference.

2. That test case is captures a real issue, due to the same hash. Now I see what you were trying to point on. Thanks for the reporting, I'm gonna fix it.